lib/Dialect/HW/Transforms/HWBypassInnerSymbols.cpp

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This pass moves inner symbols from ports to wires, then bypasses wire
+//  operations with inner symbols by replacing uses with their inputs while
+//  keeping the wire to preserve the symbol. This enables optimizations to
+//  cross symbol boundaries while maintaining symbol references.
+//
+//  Note: This transformation assumes that values associated with inner
+//  symbols are not mutated through inner symbols (e.g. force). This assumption
+//  may not hold in design verification contexts, but is safe in synthesis.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "circt/Dialect/HW/PortImplementation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+
+namespace circt {
+namespace hw {
+#define GEN_PASS_DEF_HWBYPASSINNERSYMBOLS
+#include "circt/Dialect/HW/Passes.h.inc"
+} // namespace hw
+} // namespace circt
+
+using namespace circt;
+using namespace circt::hw;
+
+namespace {
+struct HWBypassInnerSymbolsPass
+    : public impl::HWBypassInnerSymbolsBase<HWBypassInnerSymbolsPass> {
+  void runOnOperation() override;
+};
+
+/// Pattern to bypass wire operations with inner symbols.
+struct BypassWireWithInnerSym : public OpRewritePattern<WireOp> {
+  using OpRewritePattern<WireOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(WireOp wire,
+                                PatternRewriter &rewriter) const override {
+    // Only bypass wires that have an inner symbol or use
+    if (!wire.getInnerSymAttr() || wire.use_empty())
+      return failure();
+
+    // Replace all uses of the wire with its input
+    rewriter.modifyOpInPlace(wire, [&] {
+      wire->replaceAllUsesWith(ArrayRef<Value>{wire.getInput()});
+    });
+    return success();
+  }
+};
+} // namespace
+
+void HWBypassInnerSymbolsPass::runOnOperation() {
+  auto module = getOperation();
+
+  // First move the inner symbol from the port to allow constprop and other
+  // optimizations to cross the boundary.
```

Comment:
Depending on when this is used, this will break a property sometimes relied on re:paths (OM / Properties)-- that if you're tracking a port the result will always be a /port/ specifically and never an equivalent wire.

IIRC this is already broken in some cases involving output ports that are read from, since that can't be done in HW, so we lower those to wires and put symbol on the wire.

(it's still good that you're handling output ports here, for non-from-FIRRTL lowerings, or if that has or does change... 👍, perhaps obviously ❤️ )

So I'm not saying this as some spooky discouragement 😉 but just a heads-up 👍 .